platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/SyncUtils.java

```diff
@@ -16,34 +16,400 @@
+        for (final ShadowEvent tip : tips) {
+            if (tip.getEvent().getCreatorId().equals(selfId)) {
+                return tip;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Given a list of events we think the other node may not have, reduce that list to events that we think they do not
+     * have and that are unlikely to end up being duplicate events.
+     *
+     * <p>
+     * General principles:
+     * <ul>
+     * <li>Always send self events right away.</li>
+     * <li>Don't send non-ancestors of self events unless we've known about that event for a long time.</li>
+     * </ul>
+     *
+     * @param shadowGraph          the shadow graph
+     * @param selfId               the id of this node
+     * @param nonAncestorThreshold for each event that is not a self event and is not an ancestor of a self event, the
+     *                             amount of time the event must be known about before it is eligible to be sent
+     * @param now                  the current time
+     * @param eventsTheyNeed       the list of events we think they need
+     * @return the events that should be actually sent, will be a subset of the eventsTheyNeed list
+     */
+    @NonNull
+    public static List<EventImpl> filterLikelyDuplicates(
+            @NonNull final ShadowGraph shadowGraph,
+            @NonNull final NodeId selfId,
+            @NonNull final Duration nonAncestorThreshold,
+            @NonNull final Instant now,
+            @NonNull final List<EventImpl> eventsTheyNeed) {
+
+        final ShadowEvent latestSelfEvent = getLatestSelfEventInShadowgraph(shadowGraph, selfId);
+
+        final Set<ShadowEvent> selfEventAncestors;
+        if (latestSelfEvent == null) {
+            selfEventAncestors = Set.of();
+        } else {
+            final List<ShadowEvent> listOfLatestSelfEvent = List.of(latestSelfEvent);
+            selfEventAncestors = shadowGraph.findAncestors(listOfLatestSelfEvent, event -> true);
+        }
+
+        // Convert to a list of hashes for easy lookup.
+        final List<Hash> selfEventAncestorHashes =
+                selfEventAncestors.stream().map(ShadowEvent::getEventBaseHash).toList();
+
+        final List<EventImpl> filteredList = new ArrayList<>();
+
+        for (final EventImpl event : eventsTheyNeed) {
+            if (event.getCreatorId().equals(selfId)) {
+                // Always send self events right away.
+                filteredList.add(event);
+                continue;
+            }
+
+            final Instant eventReceivedTime = event.getBaseEvent().getTimeReceived();
+            final Duration timeKnown = Duration.between(eventReceivedTime, now);
+
+            final boolean isAncestor = selfEventAncestorHashes.contains(event.getBaseHash());
+
+            if (isAncestor) {
+                // Always send ancestors of self events right away.
+                filteredList.add(event);
```

Comment:
merge this block and the next with an `or`, since the block contents are identical, or `continue` after adding the event to the list

Reply:
done